File: src/NzbDrone.Core/History/History.cs
using System;
using System.Collections.Generic;
using NzbDrone.Core.Datastore;
using NzbDrone.Core.Languages;
using NzbDrone.Core.Movies;
using NzbDrone.Core.Qualities;

namespace NzbDrone.Core.History
{
    public class MovieHistory : ModelBase
    {
        public const string DOWNLOAD_CLIENT = "downloadClient";
        public const string MOVIE_MATCH_TYPE = "movieMatchType";
        public const string RELEASE_SOURCE = "releaseSource";

        public MovieHistory()
        {
            Data = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        }

        public int MovieId { get; set; }
        public string SourceTitle { get; set; }
        public QualityModel Quality { get; set; }
        public DateTime Date { get; set; }
        public Movie Movie { get; set; }
        public MovieHistoryEventType EventType { get; set; }
        public Dictionary<string, string> Data { get; set; }
        public List<Language> Languages { get; set; }

        public string DownloadId { get; set; }
    }

    public enum MovieHistoryEventType
    {
        Unknown = 0,
        Grabbed = 1,

        // SeriesFolderImported = 2, // deprecated
        DownloadFolderImported = 3,
        DownloadFailed = 4,

        // EpisodeFileDeleted = 5, // deprecated
        MovieFileDeleted = 6,
        MovieFolderImported = 7, // not used yet
        MovieFileRenamed = 8,
        DownloadIgnored = 9
    }
}

File: frontend/src/typings/History.ts
import Language from 'Language/Language';
import { QualityModel } from 'Quality/Quality';
import CustomFormat from './CustomFormat';

export type HistoryEventType =
  | 'grabbed'
  | 'downloadFolderImported'
  | 'downloadFailed'
  | 'movieFileDeleted'
  | 'movieFolderImported'
  | 'movieFileRenamed'
  | 'downloadIgnored';

export interface GrabbedHistoryData {
  indexer: string;
  nzbInfoUrl: string;
  releaseGroup: string;
  age: string;
  ageHours: string;
  ageMinutes: string;
  publishedDate: string;
  downloadClient: string;
  downloadClientName: string;
  size: string;
  downloadUrl: string;
  guid: string;
  tmdbId: string;
  imdbId: string;
  protocol: string;
  customFormatScore?: string;
  movieMatchType: string;
  releaseSource: string;
  indexerFlags: string;
}

export interface DownloadFailedHistory {
  message: string;
}

export interface DownloadFolderImportedHistory {
  customFormatScore?: string;
  droppedPath: string;
  importedPath: string;
}

export interface MovieFileDeletedHistory {
  customFormatScore?: string;
  reason: 'Manual' | 'MissingFromDisk' | 'Upgrade';
}

export interface MovieFileRenamedHistory {
  sourcePath: string;
  sourceRelativePath: string;
  path: string;
  relativePath: string;
}

export interface DownloadIgnoredHistory {
  message: string;
}

export type HistoryData =
  | GrabbedHistoryData
  | DownloadFailedHistory
  | DownloadFolderImportedHistory
  | MovieFileDeletedHistory
  | MovieFileRenamedHistory
  | DownloadIgnoredHistory;

export default interface History {
  movieId: number;
  sourceTitle: string;
  languages: Language[];
  quality: QualityModel;
  customFormats: CustomFormat[];
  customFormatScore: number;
  qualityCutoffNotMet: boolean;
  date: string;
  downloadId: string;
  eventType: HistoryEventType;
  data: HistoryData;
  id: number;
}
